### PR TITLE
[Docker] Move local image to CUDA 12.1. Add option for CPU local image.

### PR DESCRIPTION
**Pull request by thvasilo**
*Issue #, if available:*

*Description of changes:*

Motivation: we want to be able to run local CPU clusters for cost-efficient partition and inference. Also the 11.7 Nvidia image will be EOL and removed by Nvidia in July 2024.

* Move local image to CUDA 12.1, ~~DGL 2.1.0~~
* Add the option to create CPU images for local clusters, allowing us to run partition/inference in CPU local clusters.
* Also changes tagging for SageMaker image to include image type as tag suffix, cpu/gpu.

### Testing

Created local CPU and GPU clusters on EC2, run example ACM job.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docker/Dockerfile.local [hunk 1]
```diff
@@ -1,34 +1,66 @@
...
+    python3-dev \
+    python3-pip \
+    && rm -rf /var/lib/apt/lists/*
 
-# Install Pytorch
-RUN pip3 install networkx==3.1 pydantic
-RUN pip3 install torch==2.1.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
+RUN pip3 install torch==2.1.2 --index-url https://download.pytorch.org/whl/cpu
+
+RUN pip3 install networkx==3.1 pydantic dgl==2.1.0 \
```
Comment (jalencato) [inline docker/Dockerfile.local:30]: I have met a problem for this, check issue here: https://github.com/awslabs/graphstorm/issues/809 It is the problem with DGL with version >= 2.0.0, but it has no problem with DGL == 1.1.3

Comment (jalencato) [reply docker/Dockerfile.local:30]: I have also created an issue on DGL repo: https://github.com/dmlc/dgl/issues/7333, we can use DGL 1.1.3 version first.

Comment (thvasilo) [reply docker/Dockerfile.local:30]: Ok I'll try to reproduce this and can keep using DGL 1.1.x for now.

File: docs/source/scale/sagemaker.rst [hunk 2]
```diff
@@ -54,9 +54,9 @@ The ``build_docker_sagemaker.sh`` script takes four arguments:
 .. warning::
     In order to upload the GraphStorm SageMaker Docker image to Amazon ECR, users need to define the <DOCKER_NAME> to include the ECR URI string, **<AWS_ACCOUNT_ID>.dkr.ecr.<REGION>.amazonaws.com/**, e.g., ``888888888888.dkr.ecr.us-east-1.amazonaws.com/graphstorm``.
 
-4. **DOCKER_TAG** (optional), is the assigned tag name of the to-be built Docker image. Default is ``sm``.
+4. **DOCKER_TAG** (optional), is the assigned tag name of the to-be built Docker image. Default is ``sm-<cpu/gpu>``.
```
Comment (classicsong) [inline docs/source/scale/sagemaker.rst:57]: `sm-<cpu/gpu>` is a little bit confusing. Let's use `sm-cpu` and `sm-gpu` for cpu image and gpu image respectively.

File: docs/source/install/env-setup.rst [hunk 3]
```diff
@@ -121,13 +121,14 @@ Please use the following command to build a Docker image from source:
...
+    bash /path-to-graphstorm/docker/build_docker_oss4local.sh /path-to-graphstorm/ docker-name docker-tag device
 
-There are three arguments of the ``build_docker_oss4local.sh``:
+There are four positional arguments for ``build_docker_oss4local.sh``:
 
 1. **path-to-graphstorm** (**required**), is the absolute path of the "graphstorm" folder, where you cloned the GraphStorm source code. For example, the path could be ``/code/graphstorm``.
 2. **docker-name** (optional), is the assigned name of the to be built Docker image. Default is ``graphstorm``.
-3. **docker-tag** (optional), is the assigned tag name of the to be built docker image. Default is ``local``.
+3. **docker-tag** (optional), is the assigned tag name of the to be built docker image. Default is ``local-<device>``.
+3. **device** (optional), is the intended device for the docker image. Default is ``gpu``, can also build a ``cpu`` image.
```
Comment (classicsong) [inline docs/source/install/env-setup.rst:131]:
```suggestion
4. **device** (optional), is the intended device for the docker image. Default is ``gpu``, can also build a ``cpu`` image.
```

File: docker/build_docker_sagemaker.sh [hunk 4]
```diff
@@ -39,7 +39,7 @@ cp -r "${GSF_HOME}/sagemaker" code/graphstorm/sagemaker
 cp -r "${GSF_HOME}/docker/sagemaker/build_artifacts" build_artifacts
 
 # Build OSS docker for EC2 instances that an pull ECR docker images
-DOCKER_FULLNAME="${IMAGE_NAME}:${TAG}"
+DOCKER_FULLNAME="${IMAGE_NAME}:${TAG}-${IMAGE_TYPE}"
```
Comment (classicsong) [inline docker/build_docker_sagemaker.sh:42]: DEVICE_TYPE?

Comment (thvasilo) [reply docker/build_docker_sagemaker.sh:42]: We're using IMAGE_TYPE right now see https://github.com/awslabs/graphstorm/blob/b0a8c7750896c8aa0866be2318c25fabbf850b48/docker/build_docker_sagemaker.sh#L13-L18

Agree that DEVICE_TYPE is a better name, I can change it to that if you think we should.

File: docker/Dockerfile.local (hunk 1 above)
Comment (thvasilo) [reply docker/Dockerfile.local:30]: I tried to repro this inside a CPU Docker container but haven't encountered this error. Do you know what yuo particular execution environment was?

Comment (jalencato) [reply docker/Dockerfile.local:30]: I think you can try that in any GPU docker container and you will meet that problem.

Comment (thvasilo) [reply docker/Dockerfile.local:30]: Ah only happens in GPU envs, I got the error when running on a G5 with the GPU image.

**Comment by thvasilo** [comment]
Moved back to using DGL 1.1.3 and updated the Dockerfile structure to use a common source.

Could you take another look @classicsong @jalencato ? Tested on single instance and multi-instance training, with the repro example that @jalencato developed for the DGL issue.

File: docker/Dockerfile.local [hunk 5]
```diff
@@ -1,34 +1,76 @@
...
-RUN apt-get install -y cython3 libicu-dev
-RUN pip3 install h5py psutil
+FROM base as base-gpu
 
-RUN apt-get install -y unzip
+# Install torch, DGL, and GSF deps that require torch
+RUN pip install \
+    dgl==${DGL_VERSION}+cu121 \
+    ogb==${OGB_VERSION} \
+    torch==${TORCH_VERSION} \
```
Comment (jalencato) [inline docker/Dockerfile.local:61]: Why do we not specify the torch version with CUDA version any more here? I see we are doing so for dgl library installation. It is a good practice to keep them same.

File: docker/Dockerfile.local [hunk 6]
```diff
@@ -1,34 +1,76 @@
...
+    -f https://data.dgl.ai/wheels/cu121/repo.html \
+    && rm -rf /root/.cache
+
+FROM base-${DEVICE} as runtime
+
+ENV PYTHONPATH="/root/dgl/tools/:${PYTHONPATH}"
 
 # Download DGL source code
-RUN cd /root; git clone --recursive https://github.com/dmlc/dgl.git
+RUN cd /root; git clone --single-branch https://github.com/dmlc/dgl.git
```
Comment (jalencato) [inline docker/Dockerfile.local:71]: We can add ```ENV PYTHONPATH="/root/dgl/tools:$PYTHONPATH"``` as well here.

**Review by jalencato — COMMENTED** [review]
Some minor comments. Overall looks good to me.

File: docker/Dockerfile.local (hunk 5 above)
Comment (thvasilo) [reply docker/Dockerfile.local:61]: I don't think it's possible to specify CUDA 12.1 right now for PT, however it's the default. See https://pytorch.org/get-started/locally/

```
ERROR: Could not find a version that satisfies the requirement torch==2.1.2+cu121 (from versions: 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0, 2.0.1, 2.1.0, 2.1.1, 2.1.2, 2.2.0, 2.2.1, 2.2.2)
#8 5.320 ERROR: No matching distribution found for torch==2.1.2+cu121
```

For 11.8 we would need to use `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118`

File: docker/Dockerfile.local (hunk 6 above)
Comment (thvasilo) [reply docker/Dockerfile.local:71]: It's in line 68.

File: docker/Dockerfile.local (hunk 5 above)
Comment (jalencato) [reply docker/Dockerfile.local:61]: I see. Then we can leave if for now.

File: docker/build_docker_sagemaker.sh (hunk 4 above)
Comment (classicsong) [reply docker/build_docker_sagemaker.sh:42]: Let's use DEVICE_TYPE

**Review by classicsong — APPROVED** [review]
LGTM

**Review by jalencato — APPROVED** [review]
LGTM DGL said they have fixed the bug and they will include that change in 2.2 version. We can move to 2.2 later in May

**Pull request merged**